### PR TITLE
Sort modules before to register the strings

### PR DESCRIPTION
**Pull request by strategio**
I struggled a little to find the smartest logic to sort the modules, and also to implement the tests, but it looks fine now.

The data used for the test was taken from my test page (just removed the unnecessary stuff) => see https://github.com/OnTheGoSystems/wpml-page-builders-beaver-builder/compare/wpmlcore-6331?expand=1#diff-77dbb17f691cc1dc504c62173c5c2345R128.

https://onthegosystems.myjetbrains.com/youtrack/issue/wpmlcore-6331

### CODE REVIEW

**Comment by strategio** [comment]
@OnTheGoSystems/wpml

**Review by kagg-design — DISMISSED** [review]
+1 @strategio

File: src/class-wpml-beaver-builder-register-strings.php [hunk 1]
```diff
@@ -20,4 +21,47 @@ protected function register_strings_for_modules( array $data_array, array $packa
...
+	}
+
+	/**
+	 * @param stdClass $a
+	 * @param stdClass $b
+	 *
+	 * @return int
+	 */
+	private function sort_modules_by_position_only( stdClass $a, stdClass $b ) {
+		return ( (int) $a->position < (int) $b->position ) ? -1 : 1;
```
Comment (andreasciamanna) [inline src/class-wpml-beaver-builder-register-strings.php:85]: @strategio You can simplify it with `return (int) $a->position - (int) $b->position`

Comment (andreasciamanna) [reply src/class-wpml-beaver-builder-register-strings.php:85]: See here: https://gist.github.com/andreasciamanna/4031fa35830af267d7986494cd3d210e

File: src/class-wpml-beaver-builder-register-strings.php [hunk 2]
```diff
@@ -20,4 +21,47 @@ protected function register_strings_for_modules( array $data_array, array $packa
...
+	}
+
+	/**
+	 * @param array        $all_modules
+	 * @param string|null  $parent_hash
+	 * @param array        $sorted_modules
+	 *
+	 * @return array
+	 */
+	private function sort_modules_by_parent_and_child( array $all_modules, $parent_hash = null, array $sorted_modules = array() ){
```
Comment (andreasciamanna) [inline src/class-wpml-beaver-builder-register-strings.php:65]: @strategio `$all_modules` makes me think the input has something special compared to `$modules` used in `sort_modules_before_string_registration`, but instead, it's the same data: can't you use the same name everywhere?

Comment (andreasciamanna) [reply src/class-wpml-beaver-builder-register-strings.php:65]: Oh, now I think I get it.

These are the "flat" representation of the modules and what `sort_modules_by_parent_and_child` seem to do is creating some hierarchy sorted version of the array. Am I right?

I think a description of this method, in the PHPDoc, is badly needed.

**Review by andreasciamanna — CHANGES_REQUESTED** [review]

File: src/class-wpml-beaver-builder-register-strings.php (hunk 2 above)
Comment (strategio) [reply src/class-wpml-beaver-builder-register-strings.php:65]: Yes @andreasciamanna, we get a flat tree of modules. The unit tests demonstrates this https://github.com/OnTheGoSystems/wpml-page-builders-beaver-builder/pull/16/files#diff-77dbb17f691cc1dc504c62173c5c2345R128. I added a comment to make this clearer.

File: src/class-wpml-beaver-builder-register-strings.php (hunk 1 above)
Comment (strategio) [reply src/class-wpml-beaver-builder-register-strings.php:85]: @andreasciamanna, as I mentionned in the snippet, this does not behave exactly the same. Anyway I tested your suggestion and the related test is failing, so I prefer to keep it as is.

**Review by andreasciamanna — DISMISSED** [review]

**Pull request merged**